python/src/aiconfig/editor/client/src/components/EditorContainer.tsx

```diff
@@ -665,9 +669,45 @@ export default function EditorContainer({
     return () => window.removeEventListener("keydown", saveHandler);
   }, [onSave]);
 
+  // Server heartbeat, check every 3s to show error if server is down
+  const getServerStatusCallback = callbacks.getServerStatus;
+  useEffect(() => {
+    if (!getServerStatusCallback) {
+      return;
+    }
+
+    const interval = setInterval(async () => {
+      try {
+        const res = await getServerStatusCallback();
+        setServerStatus(res.status);
+      } catch (err: unknown) {
+        setServerStatus("ERROR");
+      }
+    }, 3000);
+
+    return () => clearInterval(interval);
+  }, [getServerStatusCallback]);
+
   return (
     <AIConfigContext.Provider value={contextValue}>
       <Notifications />
+      {serverStatus !== "OK" && (
+        <>
+          {/* // Simple placeholder block div to make sure the banner does not overlap page contents until scrolling past its height */}
+          <div style={{ height: "100px" }} />
+          <Alert
+            color="red"
+            title="Server Connection Error"
+            w="100%"
+            style={{ position: "fixed", top: 0, zIndex: 999 }}
+          >
+            <Text>
+              There is a problem with the editor server connection. Please copy
+              important changes somewhere safe and try restarting the editor.
+            </Text>
```

Comment:
What do you think of allowing people to save the json from the frontend state? Or at least a "copy" button for it?

Reply:
Can be in follow-up diff

Reply:
Added a copy button here; client-based save can be a follow-up if it will work, just need to dedicate time to looking into it

Reply:
<img width="1824" alt="Screenshot 2024-01-05 at 2 26 46 PM" src="https://github.com/lastmile-ai/aiconfig/assets/5060851/cf98e152-73c8-4763-92ac-958c4f25a7cc">